hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -364,31 +372,288 @@ static bool isSecureOtpMode(uint32_t normalContent) {
+            cycle = (SPIC->fbaudr);
+        } else {
+            cycle = (SPIC->baudr);
+        }
+        cycle = (cycle * dummyCycle * 2) + RTK_SPIC_READ_TUNING_DUMMY_CYCLE;
+        if (cycle <= 0x10000) {
+            RTK_SPIC_BITS_SET_VAL(SPIC->auto_length, RTK_SPIC_BIT_OFFSET_AUTO_LEN_DUM, cycle, RTK_SPIC_BIT_WIDTH_AUTO_LEN_DUM);
+        }
+    } else {
+        /* Set tuning dummy cycles */
+        RTK_SPIC_BITS_SET_VAL(SPIC->auto_length, RTK_SPIC_BIT_OFFSET_AUTO_LEN_DUM, RTK_SPIC_READ_TUNING_DUMMY_CYCLE, RTK_SPIC_BIT_WIDTH_AUTO_LEN_DUM);
+    }
+
+    /* set ctrlr0: RX mode, data_ch, addr_ch */
+    SPIC->ssienr = 0;
+    SPIC->ctrlr0 = (SPIC->ctrlr0 | 0x0300);
+    SPIC->ctrlr0 = (SPIC->ctrlr0 & 0xfff0ffff);
+
+    /* set flash cmd + addr and write to fifo */
+    SPIC->dr[0].word = RTK_SPIC_CMD_ADDR_FORMAT(cmd, addr);
+}
+
+static int gd25ReadSecurityRegisters(uint8_t otpPageIdx, uint32_t addr, uint8_t* buf, size_t size) {
+    CHECK_TRUE(otpPageIdx < HAL_QSPI_FLASH_OTP_SECTOR_COUNT_GD25, SYSTEM_ERROR_INVALID_ARGUMENT);
+    CHECK_TRUE(addr + size <= HAL_QSPI_FLASH_OTP_SIZE_GD25, SYSTEM_ERROR_INVALID_ARGUMENT);
+
+    uint32_t address = ((otpPageIdx + 1) << 12) + addr;
+    uint32_t dummyCycle = 8;
+    uint8_t cmd = GD_FLASH_CMD_RD_SEC;
+
+    uint32_t ctrl0 = SPIC->ctrlr0;
+    uint32_t autoLen = SPIC->auto_length;
+    uint32_t addrLen = SPIC->addr_length;
+    SCOPE_GUARD({
+        SPIC->ssienr = 0;
+        SPIC->ctrlr0 = ctrl0;
+        SPIC->auto_length = autoLen;
+        SPIC->addr_length = addrLen;
+    });
+
+    /* setting auto mode CS_H_WR_LEN, address length(3 byte) */
+    SPIC->auto_length = ((SPIC->auto_length & 0xfffcffff) | 0x00030000);
+    SPIC->addr_length = 3;
+
+    rtkSpiFlashSendCmdAddr(cmd, address, dummyCycle);
+
+    uint32_t cnt;
+    uint32_t data;
+    if (size <= RTK_SPIC_FIFO_HALF_SIZE) { // using RTK_SPIC_FIFO_SIZE it won't work if size >32
+        rtkSpiFlashEnableCsRead(size);
+        cnt = size / 4;
+        for (uint32_t i = 0; i < cnt; i++) {
+            data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+            memcpy(buf, &data, 4);
+            buf += 4;
+        }
+        cnt = size % 4;
+        if (cnt > 0) {
+            data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+            memcpy(buf, &data, cnt);
+            buf += cnt;
+        }
+    } else {
+        rtkSpiFlashEnableCsRead(RTK_SPIC_FIFO_HALF_SIZE);
+        cnt = RTK_SPIC_FIFO_HALF_SIZE / 4;
+        for (uint32_t i = 0; i < cnt; i++) {
+            data = rtkSpiFlashGetDataReg(0, RTK_SPIC_DATA_WORD);
+            memcpy(buf, &data, 4);
+            buf += 4;
+        }
+        /* t is transfer data times,
+         * t = (size - first xfer data) / fifo size
+         */
+        uint32_t t = (size - RTK_SPIC_FIFO_HALF_SIZE) / RTK_SPIC_FIFO_HALF_SIZE;
+        address = address + RTK_SPIC_FIFO_HALF_SIZE;
+        while (t > 0) {
+            rtkSpiFlashSendCmdAddr(cmd, address, dummyCycle);
```

Comment:
This looks like when we want to read more than 32 bytes of data in total, we need to do it with repeated 32 byte read commands (ie we have to re-send the read command `0x48` and the incrementing address)?
Seems strange this is how we would have to do this since the data sheet implies that we can continue reading and the address should just increment